winpty/src/windows.rs

```diff
@@ -40,61 +43,44 @@ bitflags!(
 );
 
 #[derive(Debug)]
-pub struct Err<'a> {
-    ptr: &'a mut winpty_error_t,
+pub struct Err {
     code: u32,
     message: String,
 }
 
 // Check to see whether winpty gave us an error
-fn check_err<'a>(e: *mut winpty_error_t) -> Option<Err<'a>> {
-    let err = unsafe {
+fn check_err(e: *mut winpty_error_t) -> Option<Err> {
+    let (code, message);
+    unsafe {
         let raw = winpty_error_msg(e);
-        Err {
-            ptr: &mut *e,
-            code: winpty_error_code(e),
-            message: String::from_utf16_lossy(std::slice::from_raw_parts(raw, wcslen(raw))),
-        }
+        code = winpty_error_code(e);
+        message = String::from_utf16_lossy(std::slice::from_raw_parts(raw, wcslen(raw)));
+        winpty_error_free(e);
     };
-    if err.code == 0 {
-        None
-    } else {
-        Some(err)
-    }
+    Some(Err { code, message }).filter(|e| e.code != 0)
```

Comment:
I think this is too 'smart'. The if/else might be a bit longer, but I think it's easier to read.

I also don't think it's nice to define a variable and then assign it in an unsafe scope, instead of returning the variables from that scope, because it makes it hard in longer function bodies to understand where exactly variables are initialized.

So `let (code, message) = unsafe { ... }` would be more clear and should be the way we do things by default.

This is probably how I would write this function:
```rust
unsafe {
    let code = match winpty_error_code(e) {
        0 => return None,
        code => code,
    };
    let raw_msg = winpty_error_msg(e);
    let message = String::from_utf16_lossy(std::slice::from_raw_parts(raw, wcslen(raw_msg)));
    Some(Err { code, message })
}
```

Reply:
As a fan of the suckless style, I agree. Unfortunately the ordering is largely dictated by the need to call winpty_error_free(e) before returning, but after copying the referenced string. Obviously code clarity trumps keeping a match statement out of the unsafe block.